compiler/dcalc/typing.ml

```diff
@@ -204,375 +205,475 @@ let op_type (op : A.operator Pos.marked) : typ Pos.marked UnionFind.elem =
   | Binop (Mult KDate) | Binop (Div KDate) | Unop (Minus KDate) ->
     Errors.raise_spanned_error pos "This operator is not available!"
 
-let rec ast_to_typ (ty : A.typ) : typ =
-  match ty with
-  | A.TLit l -> TLit l
-  | A.TArrow (t1, t2) ->
-    TArrow
-      ( UnionFind.make (Pos.map_under_mark ast_to_typ t1),
-        UnionFind.make (Pos.map_under_mark ast_to_typ t2) )
-  | A.TTuple (ts, s) ->
-    TTuple
-      ( List.map (fun t -> UnionFind.make (Pos.map_under_mark ast_to_typ t)) ts,
-        s )
-  | A.TEnum (ts, e) ->
-    TEnum
-      ( List.map (fun t -> UnionFind.make (Pos.map_under_mark ast_to_typ t)) ts,
-        e )
-  | A.TArray t -> TArray (UnionFind.make (Pos.map_under_mark ast_to_typ t))
-  | A.TAny -> TAny (Any.fresh ())
+(** {1 Double-directed typing} *)
 
-let rec typ_to_ast (ty : typ Pos.marked UnionFind.elem) : A.typ Pos.marked =
-  Pos.map_under_mark
-    (fun ty ->
-      match ty with
-      | TLit l -> A.TLit l
-      | TTuple (ts, s) -> A.TTuple (List.map typ_to_ast ts, s)
-      | TEnum (ts, e) -> A.TEnum (List.map typ_to_ast ts, e)
-      | TArrow (t1, t2) -> A.TArrow (typ_to_ast t1, typ_to_ast t2)
-      | TAny _ -> A.TAny
-      | TArray t1 -> A.TArray (typ_to_ast t1))
-    (UnionFind.get (UnionFind.find ty))
+type env = typ Marked.pos UnionFind.elem A.VarMap.t
 
-(** {1 Double-directed typing} *)
+let add_pos e ty = Marked.mark (A.pos e) ty
+let ty (_, A.Typed { ty; _ }) = ty
+let ( let+ ) x f = Bindlib.box_apply f x
+let ( and+ ) x1 x2 = Bindlib.box_pair x1 x2
 
-type env = typ Pos.marked UnionFind.elem A.VarMap.t
+let bmap f es =
```

Comment:
Ah yeah this would deserve a litlle comment!
It's basically a combination of `box_apply` and `List.map` ; I guess it would be equivalent to `List.map |> Bindlib.box_list`…